be/tx_bulk.c

```diff
@@ -592,31 +593,27 @@ M0_INTERNAL bool m0_be_tx_bulk_put(struct m0_be_tx_bulk   *tb,
 		.bbd_payload_size = payload_credit,
 	};
 	bool                    put_fail;
-	bool                    done;
+	M0_ENTRY("tb=%p", tb);
 
 	M0_PRE(partition < tb->btb_cfg.tbc_partitions_nr);
 
 	be_tx_bulk_lock(tb);
 	put_fail = tb->btb_tx_open_failed;
-	done     = tb->btb_done;
-	be_tx_bulk_unlock(tb);
-	M0_ASSERT(!done);
 
 	if (put_fail) {
+		be_tx_bulk_unlock(tb);
 		m0_be_op_active(op);
 		m0_be_op_done(op);
-		return false;
+		return M0_RC(false);
 	}
 
 	m0_be_queue_lock(&tb->btb_q[partition]);
 	M0_BE_QUEUE_PUT(&tb->btb_q[partition], op, &data);
 	m0_be_queue_unlock(&tb->btb_q[partition]);
 
-	be_tx_bulk_lock(tb);
```

Comment:
I think this was never needed as I do not see a use of this done variable. We could just remove these 4 lines and this should be enough.

Reply:
The assert() may also fail on line 603.
That's why I remove the checking of "tb->btb_done" flag.

Reply:
Further more, we can not release the "be_tx_bulk_lock(tb);",
because if:
* release the lock at LINE 602,
* BE TX worker thread might acquire the lock and run. And then it may set the "tb->btb_tx_open_failed" and the "tb->btb_done". 
* If the  ""tb->btb_done" is set, all the worker threads are done.
* The following M0_BE_QUEUE_PUT() will put an op into the queue, but the queue will not be processed. This will lead panic().


Just for your information, the following patch (based on this branch) will cause panic:
```
apr-u520428l001:cortx-motr huanghua$ git diff
diff --git a/be/tx_bulk.c b/be/tx_bulk.c
index 35c45df..8e9bbd9 100644
--- a/be/tx_bulk.c
+++ b/be/tx_bulk.c
@@ -599,9 +599,9 @@ M0_INTERNAL bool m0_be_tx_bulk_put(struct m0_be_tx_bulk   *tb,
 
        be_tx_bulk_lock(tb);
        put_fail = tb->btb_tx_open_failed;
+       be_tx_bulk_unlock(tb);
 
        if (put_fail) {
-               be_tx_bulk_unlock(tb);
                m0_be_op_active(op);
                m0_be_op_done(op);
                return M0_RC(false);
@@ -611,8 +611,6 @@ M0_INTERNAL bool m0_be_tx_bulk_put(struct m0_be_tx_bulk   *tb,
        M0_BE_QUEUE_PUT(&tb->btb_q[partition], op, &data);
        m0_be_queue_unlock(&tb->btb_q[partition]);
 
-       be_tx_bulk_unlock(tb);
-
        return M0_RC(true);
 }
 
```

Panic like this:
```
----- run_ut -t be-ut:tx_bulk-error_reg,be-ut:tx_bulk-error_payload -----
be-ut 
  tx_bulk-error_reg  motr[76397]:  5be0  FATAL  [lib/assert.c:50:m0_panic]  panic: bq->bq_enqueued == bq->bq_dequeued at m0_be_queue_fini() (be/queue.c:156)  [git: sage-base-1.0-163-g1ef98c2-dirty] /var/motr/m0ut/m0trace.76397
Motr panic: bq->bq_enqueued == bq->bq_dequeued at m0_be_queue_fini() be/queue.c:156 (errno: 4) (last failed: none) [git: sage-base-1.0-163-g1ef98c2-dirty] pid: 76397  /var/motr/m0ut/m0trace.76397
Motr panic reason: bq=(queue=0x1404d90 bq_enqueued=7 bq_dequeued=6 bq_the_end=1)
/work/cortx-motr/motr/.libs/libmotr.so.1(m0_arch_backtrace+0x20)[0x7f5d3b0f7800]
/work/cortx-motr/motr/.libs/libmotr.so.1(m0_arch_panic+0xe6)[0x7f5d3b0f79b6]
/work/cortx-motr/motr/.libs/libmotr.so.1(+0x358a84)[0x7f5d3b0e6a84]
/work/cortx-motr/motr/.libs/libmotr.so.1(+0x2b5404)[0x7f5d3b043404]
/work/cortx-motr/motr/.libs/libmotr.so.1(m0_be_tx_bulk_fini+0x84)[0x7f5d3b049734]
/work/cortx-motr/ut/.libs/libmotr-ut.so.0(+0x132247)[0x7f5d3c6c8247]
/work/cortx-motr/ut/.libs/libmotr-ut.so.0(+0x13273e)[0x7f5d3c6c873e]
/work/cortx-motr/ut/.libs/libmotr-ut.so.0(m0_be_ut_tx_bulk_error_reg+0x4a)[0x7f5d3c6c8aba]
/work/cortx-motr/ut/.libs/libmotr-ut.so.0(m0_ut_run+0x25e)[0x7f5d3c80798e]
/work/cortx-motr/ut/.libs/lt-m0ut(main+0x1156)[0x404466]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f5d3988c505]
/work/cortx-motr/ut/.libs/lt-m0ut[0x4045ba]
Motr pid 76397 waits for debugger.
```

Reply:
I think you are right. Let us go with your code change.